// @earendil-works/pi-ai/dist/providers/register-builtins.js
var __rewriteRelativeImportExtension = (this && this.__rewriteRelativeImportExtension) || function (path, preserveJsx) {
    if (typeof path === "string" && /^\.\.?\//.test(path)) {
        return path.replace(/\.(tsx)$|((?:\.d)?)((?:\.[^./]+?)?)\.([cm]?)ts$/i, function (m, tsx, d, ext, cm) {
            return tsx ? preserveJsx ? ".jsx" : ".js" : d && (!ext || !cm) ? m : (d + ext + "." + cm.toLowerCase() + "js");
        });
    }
    return path;
};
import { clearApiProviders, getApiProvider, registerApiProvider } from "../api-registry.js";
import { getImagesApiProvider, registerImagesApiProvider } from "../images-api-registry.js";
import { AssistantMessageEventStream } from "../utils/event-stream.js";
function createLazyLoadErrorImages(model, error) {
    return {
        api: model.api,
        provider: model.provider,
        model: model.id,
        output: [],
        stopReason: "error",
        errorMessage: error instanceof Error ? error.message : String(error),
        timestamp: Date.now(),
    };
}
function createLazyImagesApiProvider(api, loadModule) {
    return {
        api,
        generateImages: async (model, context, options) => {
            try {
                const module = await loadModule();
                module.register();
                const provider = getImagesApiProvider(api);
                if (!provider) {
                    throw new Error(`No API provider registered for api: ${api}`);
                }
                return await provider.generateImages(model, context, options);
            }
            catch (error) {
                return createLazyLoadErrorImages(model, error);
            }
        },
    };
}
const importNodeOnlyProvider = (specifier) => {
    const runtimeSpecifier = import.meta.url.endsWith(".js") ? specifier.replace(/\.ts$/, ".js") : specifier;
    return import(__rewriteRelativeImportExtension(runtimeSpecifier));
};
let bedrockProviderModuleOverride;
export function setBedrockProviderModule(module) {
    bedrockProviderModuleOverride = module;
}
function forwardStream(target, source) {
    (async () => {
        for await (const event of source) {
            target.push(event);
        }
        target.end();
    })();
}
function createLazyLoadErrorMessage(model, error) {
    return {
        role: "assistant",
        content: [],
        api: model.api,
        provider: model.provider,
        model: model.id,
        usage: {
            input: 0,
            output: 0,
            cacheRead: 0,
            cacheWrite: 0,
            totalTokens: 0,
            cost: { input: 0, output: 0, cacheRead: 0, cacheWrite: 0, total: 0 },
        },
        stopReason: "error",
        errorMessage: error instanceof Error ? error.message : String(error),
        timestamp: Date.now(),
    };
}
async function loadAndRegisterProvider(api, loadModule) {
    const module = await loadModule();
    module.register();
    const provider = getApiProvider(api);
    if (!provider) {
        throw new Error(`No API provider registered for api: ${api}`);
    }
    return provider;
}
function createLazyStream(api, loadModule) {
    return (model, context, options) => {
        const outer = new AssistantMessageEventStream();
        loadAndRegisterProvider(api, loadModule)
            .then((provider) => {
            const inner = provider.stream(model, context, options);
            forwardStream(outer, inner);
        })
            .catch((error) => {
            const message = createLazyLoadErrorMessage(model, error);
            outer.push({ type: "error", reason: "error", error: message });
            outer.end(message);
        });
        return outer;
    };
}
function createLazySimpleStream(api, loadModule) {
    return (model, context, options) => {
        const outer = new AssistantMessageEventStream();
        loadAndRegisterProvider(api, loadModule)
            .then((provider) => {
            const inner = provider.streamSimple(model, context, options);
            forwardStream(outer, inner);
        })
            .catch((error) => {
            const message = createLazyLoadErrorMessage(model, error);
            outer.push({ type: "error", reason: "error", error: message });
            outer.end(message);
        });
        return outer;
    };
}
function createLazyApiProvider(api, loadModule) {
    return {
        api,
        stream: createLazyStream(api, loadModule),
        streamSimple: createLazySimpleStream(api, loadModule),
    };
}
function registerBedrockProviderModule(module) {
    registerApiProvider({
        api: "bedrock-converse-stream",
        stream: module.streamBedrock,
        streamSimple: module.streamSimpleBedrock,
    });
}
function loadBedrockProviderModule() {
    const module = bedrockProviderModuleOverride;
    if (module) {
        return Promise.resolve({ register: () => registerBedrockProviderModule(module) });
    }
    return importNodeOnlyProvider("./amazon-bedrock.ts").then((provider) => provider);
}
const anthropicProvider = createLazyApiProvider("anthropic-messages", () => import("./anthropic.js"));
const azureOpenAIResponsesProvider = createLazyApiProvider("azure-openai-responses", () => import("./azure-openai-responses.js"));
const googleProvider = createLazyApiProvider("google-generative-ai", () => import("./google.js"));
const googleVertexProvider = createLazyApiProvider("google-vertex", () => import("./google-vertex.js"));
const mistralProvider = createLazyApiProvider("mistral-conversations", () => import("./mistral.js"));
const openAICodexResponsesProvider = createLazyApiProvider("openai-codex-responses", () => import("./openai-codex-responses.js"));
const openAICompletionsProvider = createLazyApiProvider("openai-completions", () => import("./openai-completions.js"));
const openAIResponsesProvider = createLazyApiProvider("openai-responses", () => import("./openai-responses.js"));
const bedrockProvider = createLazyApiProvider("bedrock-converse-stream", loadBedrockProviderModule);
const openRouterImagesProvider = createLazyImagesApiProvider("openrouter-images", () => import("./images/openrouter.js"));
export const generateImagesOpenRouter = openRouterImagesProvider.generateImages;
export const streamAnthropic = anthropicProvider.stream;
export const streamSimpleAnthropic = anthropicProvider.streamSimple;
export const streamAzureOpenAIResponses = azureOpenAIResponsesProvider.stream;
export const streamSimpleAzureOpenAIResponses = azureOpenAIResponsesProvider.streamSimple;
export const streamGoogle = googleProvider.stream;
export const streamSimpleGoogle = googleProvider.streamSimple;
export const streamGoogleVertex = googleVertexProvider.stream;
export const streamSimpleGoogleVertex = googleVertexProvider.streamSimple;
export const streamMistral = mistralProvider.stream;
export const streamSimpleMistral = mistralProvider.streamSimple;
export const streamOpenAICodexResponses = openAICodexResponsesProvider.stream;
export const streamSimpleOpenAICodexResponses = openAICodexResponsesProvider.streamSimple;
export const streamOpenAICompletions = openAICompletionsProvider.stream;
export const streamSimpleOpenAICompletions = openAICompletionsProvider.streamSimple;
export const streamOpenAIResponses = openAIResponsesProvider.stream;
export const streamSimpleOpenAIResponses = openAIResponsesProvider.streamSimple;
const registerBuiltInApiProviderFunctions = [
    () => registerApiProvider(anthropicProvider),
    () => registerApiProvider(openAICompletionsProvider),
    () => registerApiProvider(mistralProvider),
    () => registerApiProvider(openAIResponsesProvider),
    () => registerApiProvider(azureOpenAIResponsesProvider),
    () => registerApiProvider(openAICodexResponsesProvider),
    () => registerApiProvider(googleProvider),
    () => registerApiProvider(googleVertexProvider),
    () => registerApiProvider(bedrockProvider),
];
export function registerBuiltInImagesApiProviders() {
    registerImagesApiProvider(openRouterImagesProvider);
}
export function registerBuiltInApiProviders() {
    for (const register of registerBuiltInApiProviderFunctions) {
        register();
    }
}
export function resetApiProviders() {
    clearApiProviders();
    registerBuiltInApiProviders();
}
registerBuiltInApiProviders();
