// @heyhuynhgiabuu/pi-task/dist/subagent/runSdk.js
function resolveModel(ctx, requested) {
    const registry = ctx.modelRegistry;
    const available = registry?.getAll?.() ?? registry?.getAvailable?.() ?? [];
    if (requested) {
        const [provider, ...rest] = requested.split("/");
        const modelId = rest.join("/");
        const byProvider = available.find((model) => {
            return model?.provider?.id === provider && model?.id === modelId;
        });
        if (byProvider)
            return byProvider;
        const byId = available.find((model) => model?.id === requested || model?.name === requested);
        if (byId)
            return byId;
    }
    return available[0];
}
export async function runSdkSubagent(options) {
    const model = resolveModel(options.ctx, options.model ?? options.agent.model);
    if (!model) {
        throw new Error("No model available for SDK subagent execution");
    }
    const { createAgentSession, DefaultResourceLoader } = await import("@earendil-works/pi-coding-agent");
    const resourceLoader = new DefaultResourceLoader({
        cwd: options.cwd,
        systemPromptOverride: options.systemPrompt,
    });
    const { session } = await createAgentSession({
        cwd: options.cwd,
        model,
        thinkingLevel: options.thinkingLevel,
        tools: options.tools,
        excludeTools: options.excludeTools,
        resourceLoader,
    });
    await session.prompt(options.prompt);
    const sessionPath = session.sessionFile;
    const output = getLastAssistantText(session.messages);
    return { output: output.trim(), sessionPath };
}
function getLastAssistantText(messages) {
    for (let index = messages.length - 1; index >= 0; index -= 1) {
        const message = messages[index];
        if (message?.role !== "assistant")
            continue;
        const content = message.content;
        if (typeof content === "string")
            return content;
        if (Array.isArray(content)) {
            return content
                .map((part) => {
                if (typeof part === "string")
                    return part;
                if (typeof part?.text === "string")
                    return part.text;
                return "";
            })
                .filter(Boolean)
                .join("\n");
        }
    }
    return "";
}
