// @heyhuynhgiabuu/pi-task/dist/subagent/waitCompletion.js
import { readFile } from "node:fs/promises";
import { existsSync } from "node:fs";
import { join } from "node:path";
import { getLastAssistantTextFromSessionDir } from "../session-text.js";
import { paneExists } from "./tmux.js";
async function readResultFile(resultPath) {
    if (!existsSync(resultPath))
        return null;
    const text = (await readFile(resultPath, "utf-8")).trim();
    return text.length > 0 ? text : null;
}
function readSessionText(sessionDir, sessionName) {
    const sessionPath = join(sessionDir, "sessions", sessionName);
    const text = getLastAssistantTextFromSessionDir(sessionPath).trim();
    return text.length > 0 ? text : null;
}
export async function checkTaskCompletion(options) {
    const result = await readResultFile(options.resultPath);
    if (result) {
        return { status: "completed", content: result, source: "result-file" };
    }
    if (options.paneId && paneExists(options.paneId)) {
        return { status: "running", content: "", source: "pane" };
    }
    const sessionText = readSessionText(options.sessionDir, options.sessionName);
    if (sessionText) {
        return {
            status: "completed",
            content: sessionText,
            source: "session-jsonl",
        };
    }
    if (options.paneId) {
        return {
            status: "failed",
            content: "Task pane exited before producing a result or assistant response.",
            source: "pane",
        };
    }
    return { status: "running", content: "", source: "pane" };
}
export async function waitForTaskCompletion(options) {
    const timeoutMs = options.timeoutMs ?? 30 * 60 * 1000;
    const pollMs = options.pollMs ?? 1000;
    const deadline = Date.now() + timeoutMs;
    while (true) {
        if (options.signal?.aborted) {
            return {
                status: "cancelled",
                content: "Task was cancelled.",
                source: "signal",
            };
        }
        const snapshot = await checkTaskCompletion(options);
        if (snapshot.status !== "running")
            return snapshot;
        if (Date.now() >= deadline) {
            return {
                status: "timeout",
                content: `Task timed out after ${Math.round(timeoutMs / 1000)}s without producing a result.`,
                source: "timeout",
            };
        }
        await new Promise((resolve) => setTimeout(resolve, pollMs));
    }
}
